assets/model_monitoring/components/src/action_analyzer/action_analyzer_relevance_action_detector/relevance_action_detector.py

```diff
@@ -0,0 +1,606 @@
+        samples.append(sample)
+    return samples
+
+
+def write_actions(df, actions, action_output_folder, aml_deployment_id):
+    """Write the action summary and action detail files."""
+    local_path = str(uuid.uuid4())
+    action_summary = {}
+    for index_id, action in actions.items():
+        action_bad_df = df[(df[INDEX_ID_COLUMN] == index_id) & (df[SPAN_ID_COLUMN].isin(action.query_ids))]
+        action_good_df = df[(df[INDEX_ID_COLUMN] == index_id) &
+                            (df[DOCUMENT_RELEVANCE_SCORE_COLUMN] >= GOOD_METRICS_VALUE)]
+        action_id = str(uuid.uuid4())
+        action = {
+            "ActionId": action_id,
+            "Type": action.action_type,
+            "Description": ACTION_DESCRIPTION.replace("{index_id}", index_id),
+            "ConfidenceScore": action.confidence_score,
+            "ViolatedMetrics": "Retrieval Relevance",
+            "QueryIntention": action.query_intention,
+            "CreationTime": str(datetime.datetime.now()),
+            "FilePath": os.path.join(action_output_folder, f"actions/{action_id}.json")
+        }
+        action_summary[action_id] = action
+        action_detail = copy.deepcopy(action)
+        action_detail["DeploymentId"] = aml_deployment_id
+        action_detail["RunId"] = os.environ.get("AZUREML_RUN_ID", None)
+        if is_index_asset(index_id):
+            action_detail["IndexAssetId"] = index_id
+        else:
+            action_detail["IndexName"] = index_id
+        action_detail["IndexContent"] = action_bad_df.iloc[0][INDEX_CONTENT_COLUMN]
+        action_detail["PositiveSamples"] = generate_samples(action_good_df, False)
+        action_detail["NegativeSamples"] = generate_samples(action_bad_df, True)
+        print("Writing action detail of action: ")
+        print(action_detail)
+        write_to_file(action_detail, local_path, action_id)
+    print("Writing action summary to location ", action_output_folder)
+    print(action_summary)
+    write_to_file(action_summary, local_path, "action_summary")
+    target_remote_path = os.path.join(action_output_folder, "actions")
+    amlfs_upload(local_path=local_path, remote_path=target_remote_path)
+
+
+def add_action_tag_to_run():
+    """Add action analyzer tag for the root run."""
+    root_run_id = os.environ.get("AZUREML_ROOT_RUN_ID", None)
+    print("Action generated, setting tag for pipeline run: ", root_run_id)
+    client = MlflowClient()
+    client.set_tag(root_run_id, "momo_action_analyzer_has_action", "true")
+
+
+def relevance_action_detector(df,
+                              workspace_connection_arm_id,
+                              model_deployment_name,
+                              llm_summary_enabled,
+                              action_output_folder,
+                              aml_deployment_id):
+    """Relevance action detector."""
+    request_args = get_llm_request_args(model_deployment_name)
+
+    # Expand to span level and get document relevance score
```

Comment:
e.g. create the token manager in this function, and pass it into both l 571 and l 581